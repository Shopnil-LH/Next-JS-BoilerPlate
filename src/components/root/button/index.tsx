/**
 * Export all component defined in this folder from here
 */

import { RedirectButton } from './redirect-button';
export { RedirectButton };
