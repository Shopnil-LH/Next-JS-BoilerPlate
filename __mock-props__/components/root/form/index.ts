// Internal Importstypes
import { IForm, FormElementType } from '@src/types/root';

// Internal formData
export const formData: IForm = {
  formTitle: 'Send Us Message',
  formFields: [
    {
      elementType: FormElementType.Input,
      type: 'text',
      name: 'firstName',
      placeholder: 'First Name',
      isRequired: true,
      space: 0.5,
    },
    {
      elementType: FormElementType.Input,
      type: 'text',
      name: 'lastName',
      placeholder: 'Last Name',
      isRequired: true,
      space: 0.5,
    },
    {
      elementType: FormElementType.Input,
      type: 'text',
      name: 'email',
      placeholder: 'Enter Email Address',
      isRequired: true,
      space: 1,
    },
    {
      elementType: FormElementType.Input,
      type: 'text',
      name: 'phone',
      placeholder: 'Phone number',
      isRequired: false,
      space: 1,
    },
    {
      elementType: FormElementType.TextArea,
      name: 'message',
      placeholder: 'Write Your Message Here',
      isRequired: true,
      space: 1,
      type: 'text',
    },
    {
      label: 'Terms & Condition',
      type: 'checkbox',
      elementType: FormElementType.Checkbox,
      name: 'terms',
      isRequired: true,
      space: 1,
    },
  ],
  buttonInfo: {
    buttonText: 'Send Now',
    colorScheme: 'skyblue-over-gSkyblueLight',
    size: 'btn-sm',
    'data-testid': 'submit-button',
  },
};
