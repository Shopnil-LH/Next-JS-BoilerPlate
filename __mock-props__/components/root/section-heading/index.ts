// Internal Imports
import { ISectionHeading } from '@src/types/root';

//sectionHeadingData
export const sectionHeadingData: ISectionHeading = {
  subTitle: 'Event Speakers',
  title: 'Meet Our Speakers',
  description:
    'We’re bringing together some of the best speakers in their field, and adding more as they’re confirmed.',
  isCenter: false,
  colorScheme: 'blue',
  isBottomSpacing: false,
};
