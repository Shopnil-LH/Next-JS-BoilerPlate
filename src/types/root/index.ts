import {
  colorSchemeType,
  sizeType,
  iconNameType,
  IRedirectButton,
} from './button';
import { ISectionHeading } from './section-heading';
import {
  IFormField,
  IFormValues,
  IFormFieldProp,
  IForm,
  FormElementType,
} from './form';

import {
  IToDo,
  EventType,
  IInputProps,
  ButtonProps
}  from "./toDo"

export type {
  colorSchemeType,
  sizeType,
  iconNameType,
  IRedirectButton,
  ISectionHeading,
  IFormField,
  IFormValues,
  IFormFieldProp,
  IForm,
};

export { FormElementType };

export type{
  IToDo,
  EventType,
  IInputProps,
  ButtonProps
}