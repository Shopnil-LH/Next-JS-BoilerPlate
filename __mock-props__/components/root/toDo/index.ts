import { EventType, IToDo } from "@src/types/root";

export const toDoDemoData: IToDo = {
    toDoTitle: "My Diary",
    inputFieldProps: {
        type: "text",
        placeholder: "Set Your Schedule",
        value: "mySchedule",
        name: "mySchedule",
        buttonProps: {
            label: "Add New  Schedule",
            onCLick: (e: EventType) => {alert("Clicked!!")}
        }
    },
    lists: {
        lists: ["Hello", "Bye"]
    }
}