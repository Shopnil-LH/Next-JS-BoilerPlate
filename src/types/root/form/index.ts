// External Imports
import React from 'react';

// Internal Imports
import { IRedirectButton } from '@src/types/root';

export enum FormElementType {
  Input = 'input',
  TextArea = 'textarea',
  Checkbox = 'checkbox',
  Select = 'select',
}

// IForm
export interface IForm {
  formFields: IFormFieldProp[];
  buttonInfo: IRedirectButton;
  formTitle?: string;
  privacyInfo?: IPrivacyInfo;
}

// IFormField
export interface IFormField {
  fieldProps: IFormFieldProp;
  value: string | undefined;
  onChange?: (
    e: React.ChangeEvent<HTMLTextAreaElement | HTMLInputElement | any>
  ) => void;
  formErrors: IFormValues;
}

// IFormFieldProp
export interface IFormFieldProp {
  name: string;
  space: number;
  isRequired: boolean;
  label?: string;
  elementType: FormElementType;
  type: string;
  placeholder?: string;
}

// IFormValues
export interface IFormValues {
  [key: string]: string | undefined;
}

// IPrivacyInfo
export interface IPrivacyInfo {
  link: string;
  linkText: string;
}
