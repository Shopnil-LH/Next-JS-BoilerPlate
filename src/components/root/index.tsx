import { RedirectButton } from './button';
import { SectionHeading } from './section-heading';
import { Form } from './form';
import {ToDo} from "./toDo"

export { RedirectButton, SectionHeading, Form , ToDo};
