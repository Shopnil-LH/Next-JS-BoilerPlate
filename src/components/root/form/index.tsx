// External Imports
import { useState } from 'react';
import Link from 'next/link';

// Internal Imports
import { dataValidation } from '@src/utils';
import { IFormValues, IForm } from '@src/types/root';
import { RedirectButton } from '@src/components/root';
import { FormField } from './form-field';

/**
 * @param {object} IForm
 * @returns JSX.Element
 */
export const Form = ({
  formTitle,
  formFields,
  buttonInfo,
  privacyInfo,
}: IForm) => {
  const [formValues, setFormValues] = useState<IFormValues>({});
  const [formErrors, setFormErrors] = useState<IFormValues>({});

  const handleFormChange = (
    e: React.ChangeEvent<HTMLTextAreaElement | HTMLInputElement | any>
  ) => {
    const { name, value, type } = e.target;
    if (type === 'checkbox') {
      setFormValues({
        ...formValues,
        [name]: e.target.checked,
      });
    } else {
      setFormValues({
        ...formValues,
        [name]: value,
      });
    }
  };

  // submitHandler
  const submitHandler = (
    e: React.MouseEvent<HTMLElement> | React.FormEvent<HTMLFormElement>
  ) => {
    e.preventDefault();

    const currentFormErrors = dataValidation(formFields, formValues);
    setFormErrors(currentFormErrors);

    if (Object.keys(currentFormErrors).length == 0) {
      /**
       * Todo: Update form submission method
       */
      alert(JSON.stringify(formValues));
      setFormValues({});
    }
  };

  return (
    <form data-testid="form" className={`form grid gap-4 lg:grid-cols-2`}>
      {formTitle && (
        <div className="mb-[.5rem] md:mb-[.75rem] col-start-1 col-end-3">
          <h3 className="text-[1.75rem] font-medium leading-[1.3] md:text-[3rem] md:font-bold md:leading-[1.1] capitalize text-secondary">
            {formTitle}
          </h3>
        </div>
      )}

      {formFields &&
        formFields.map((field, index) => {
          const halfWidth = 'max-sm:col-start-1 max-sm:col-end-3';
          const fullWidth = 'col-start-1 col-end-3';
          return (
            <div
              key={index}
              className={field.space === 0.5 ? halfWidth : fullWidth}
            >
              <FormField
                fieldProps={field}
                value={formValues[field.name] ? formValues[field.name] : ''}
                onChange={handleFormChange}
                formErrors={formErrors}
              />
            </div>
          );
        })}

      {privacyInfo && (
        <div className="col-start-1 col-end-3">
          <Link href={`${privacyInfo?.link}`} passHref>
            <a className="inline-flex focus:outline-0 focus:shadow-none focus-visible:outline-0 text-[.9375rem] text-secondary underline uppercase transition-colors duration-300 hover:text-secondaryLight">
              {privacyInfo?.linkText}
            </a>
          </Link>
        </div>
      )}

      {buttonInfo.buttonText && (
        <div className="col-start-1 col-end-3">
          <RedirectButton {...buttonInfo} clickHandler={submitHandler} />
        </div>
      )}
    </form>
  );
};
