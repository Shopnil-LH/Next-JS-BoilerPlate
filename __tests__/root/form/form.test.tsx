// External Imports
import '@testing-library/jest-dom';
import { render, fireEvent } from '@testing-library/react';

// Internal Imports
import { Form } from '@src/components/root';
import { formData } from '@root/__mock-props__/components/root/form';
import {
  generalInputValidationTest,
  emailInputValidationTest,
  checkBoxInputValidationTest,
  errorBorder,
  errorText,
  normalText,
  normalBorder,
} from '@root/__tests__/root/form/config';
import { IForm } from '@src/types/root/form/index';

const specialFields = ['email', 'checkbox'];

/**
 * Test if the form is rendering correctly
 */
describe('Form Render Test', () => {
  it('renders a form correctly', () => {
    const { getByTestId } = render(<Form {...formData} />);
    const form = getByTestId('form');

    expect(form).toBeInTheDocument();
    expect(form).toMatchSnapshot();
  });
});

/**
 * Test if the form fields updating correctly
 */
describe.each(formData.formFields)(
  'Form input onChange test',
  ({ type, name }) => {
    const updatedInputString = 'Test String';

    it(`${name} change event triggers correctly`, () => {
      const { getByTestId } = render(<Form {...formData} />);
      const element = getByTestId(name) as HTMLInputElement;

      if (type === 'text') {
        fireEvent.change(element, {
          target: { value: updatedInputString },
        });
        expect(element.value).toBe(updatedInputString);
      } else if (type === 'checkbox') {
        expect(element.checked).toEqual(false);
        fireEvent.click(element);
        expect(element.checked).toEqual(true);
      }
    });
  }
);

/**
 * Test input validation for general text input fields
 */
describe.each(generalInputValidationTest)(
  'General input validation Tests with value: $value',
  ({ value, expectError }) => {
    describe.each(formData.formFields)(
      'Given Form Fields',
      ({ type, name }) => {
        /**
         * We only want to test generic text inputs
         */

        const isGeneralInputField = !specialFields.some((item) =>
          [type, name].includes(item)
        );

        if (isGeneralInputField) {
          let formFields, modifiedFormData: IForm;
          beforeAll(() => {
            /**
             * In order to run validation checks for general text fields
             * we must set isRequired to true for those fields
             */
            formFields = formData.formFields.map((item) => {
              return { ...item, isRequired: true };
            });

            modifiedFormData = {
              ...formData,
              formFields,
            };
          });

          it(`Input element for ${name} should  have ${
            expectError ? 'error' : 'normal'
          } border`, () => {
            const { getByTestId } = render(<Form {...modifiedFormData} />);
            const submitButton = getByTestId(
              modifiedFormData.buttonInfo['data-testid']
            );
            const element = getByTestId(name) as HTMLInputElement;

            fireEvent.change(element, {
              target: { value: value },
            });
            fireEvent.click(submitButton);

            if (expectError) {
              expect(element).toHaveClass(errorBorder);
            } else {
              expect(element).not.toHaveClass(errorBorder);
            }
          });
        }
      }
    );
  }
);

/**
 * Test email input validation
 */
describe.each(emailInputValidationTest)(
  'Email input validation Tests with value: $value',
  ({ value, expectError }) => {
    /**
     * We only want to test for email input
     */
    let formFields, modifiedFormData: IForm;
    beforeAll(() => {
      /**
       * In order to run validation checks for general text fields
       * we must set isRequired to true for those fields
       */
      formFields = formData.formFields.map((item) => {
        return { ...item, isRequired: true };
      });

      modifiedFormData = {
        ...formData,
        formFields,
      };
    });

    it(`Input element for email should  have ${
      expectError ? 'error' : 'normal'
    } border`, () => {
      const { getByTestId } = render(<Form {...modifiedFormData} />);
      const submitButton = getByTestId(
        modifiedFormData.buttonInfo['data-testid']
      );
      const element = getByTestId('email') as HTMLInputElement;

      fireEvent.change(element, {
        target: { value: value },
      });
      fireEvent.click(submitButton);

      if (expectError) {
        expect(element).toHaveClass(errorBorder);
      } else {
        expect(element).toHaveClass(normalBorder);
      }
    });
  }
);

/**
 * Test checkbox input validation
 */
describe.each(checkBoxInputValidationTest)(
  'Checkbox validation Tests with value: $value',
  ({ value, expectError }) => {
    let formFields, modifiedFormData: IForm;
    beforeAll(() => {
      /**
       * In order to run validation checks for general text fields
       * we must set isRequired to true for those fields
       */
      formFields = formData.formFields.map((item) => {
        return { ...item, isRequired: true };
      });

      modifiedFormData = {
        ...formData,
        formFields,
      };
    });

    it(`Checkbox box and label should  have ${
      expectError ? 'error' : 'normal'
    } styles`, () => {
      const { getByTestId } = render(<Form {...modifiedFormData} />);
      const submitButton = getByTestId(
        modifiedFormData.buttonInfo['data-testid']
      );
      const checkboxInputElement = getByTestId('terms') as HTMLInputElement;
      const checkboxBox = getByTestId('checkbox-box');
      const checkboxLabel = getByTestId('checkbox-label');

      if (value) {
        fireEvent.click(checkboxInputElement);
      }
      fireEvent.click(submitButton);

      if (expectError) {
        expect(checkboxBox).toHaveClass(errorBorder);
        expect(checkboxLabel).toHaveClass(errorText);
      } else {
        expect(checkboxBox).toHaveClass('border-secondary');
        expect(checkboxLabel).toHaveClass(normalText);
      }
    });
  }
);
