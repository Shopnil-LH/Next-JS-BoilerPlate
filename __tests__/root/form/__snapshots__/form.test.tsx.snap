// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form Render Test renders a form correctly 1`] = `
<form
  class="form grid gap-4 lg:grid-cols-2"
  data-testid="form"
>
  <div
    class="mb-[.5rem] md:mb-[.75rem] col-start-1 col-end-3"
  >
    <h3
      class="text-[1.75rem] font-medium leading-[1.3] md:text-[3rem] md:font-bold md:leading-[1.1] capitalize text-secondary"
    >
      Send Us Message
    </h3>
  </div>
  <div
    class="max-sm:col-start-1 max-sm:col-end-3"
  >
    <div>
      <input
        autocomplete="Off"
        class="border-[1px] border-solid block w-full focus:outline-0 focus:shadow-none focus-visible:outline-0  placeholder:capitalize placeholder:text-blackLight text-blackLight bg-gray text-base rounded-lg p-[1rem_1.5rem] border-transparent"
        data-testid="firstName"
        id="firstName"
        name="firstName"
        placeholder="First Name*"
        type="text"
        value=""
      />
    </div>
  </div>
  <div
    class="max-sm:col-start-1 max-sm:col-end-3"
  >
    <div>
      <input
        autocomplete="Off"
        class="border-[1px] border-solid block w-full focus:outline-0 focus:shadow-none focus-visible:outline-0  placeholder:capitalize placeholder:text-blackLight text-blackLight bg-gray text-base rounded-lg p-[1rem_1.5rem] border-transparent"
        data-testid="lastName"
        id="lastName"
        name="lastName"
        placeholder="Last Name*"
        type="text"
        value=""
      />
    </div>
  </div>
  <div
    class="col-start-1 col-end-3"
  >
    <div>
      <input
        autocomplete="Off"
        class="border-[1px] border-solid block w-full focus:outline-0 focus:shadow-none focus-visible:outline-0  placeholder:capitalize placeholder:text-blackLight text-blackLight bg-gray text-base rounded-lg p-[1rem_1.5rem] border-transparent"
        data-testid="email"
        id="email"
        name="email"
        placeholder="Enter Email Address*"
        type="text"
        value=""
      />
    </div>
  </div>
  <div
    class="col-start-1 col-end-3"
  >
    <div>
      <input
        autocomplete="Off"
        class="border-[1px] border-solid block w-full focus:outline-0 focus:shadow-none focus-visible:outline-0  placeholder:capitalize placeholder:text-blackLight text-blackLight bg-gray text-base rounded-lg p-[1rem_1.5rem] border-transparent"
        data-testid="phone"
        id="phone"
        name="phone"
        placeholder="Phone number"
        type="text"
        value=""
      />
    </div>
  </div>
  <div
    class="col-start-1 col-end-3"
  >
    <div>
      <textarea
        autocomplete="Off"
        class="border-[1px] border-solid block w-full focus:outline-0 focus:shadow-none focus-visible:outline-0  placeholder:capitalize placeholder:text-blackLight text-blackLight bg-gray text-base rounded-lg p-[1rem_1.5rem] border-transparent resize-none min-h-[130px]"
        data-testid="message"
        id="message"
        name="message"
        placeholder="Write Your Message Here*"
      />
    </div>
  </div>
  <div
    class="col-start-1 col-end-3"
  >
    <div
      class="inline-flex items-center flex-wrap gap-2"
    >
      <input
        autocomplete="Off"
        class="appearance-none hidden"
        data-testid="terms"
        id="terms"
        name="terms"
        placeholder="undefined*"
        type="checkbox"
        value=""
      />
      <label
        class="text-base font-medium inline-flex items-center flex-wrap gap-[.625rem] cursor-pointer"
        for="terms"
      >
        <span
          class="relative w-[18px] h-[18px] border border-solid  rounded-sm inline-flex items-center after:content-[''] after:absolute after:w-[12px] after:h-[12px] after:bg-secondary after:rounded-sm after:translate-x-[-50%] after:translate-y-[-50%] after:left-[50%] after:top-[50%] after:invisible border-[#cbcbcb]"
          data-testid="checkbox-box"
        />
        <span
          class="inline-block relative top-[2px]  text-blackLight"
          data-testid="checkbox-label"
        >
          Terms & Condition
        </span>
      </label>
    </div>
  </div>
  <div
    class="col-start-1 col-end-3"
  >
    <a
      class="bs-button skyblue-over-gSkyblueLight
        btn-sm  group relative cursor-pointer overflow-hidden inline-flex gap-[1rem] items-center justify-center backdrop-blur-[2px] capitalize font-medium rounded-[8px] transition-colors duration-[400ms]"
      data-testid="submit-button"
    >
      <span
        class="bs-btn-text"
      >
        Send Now
      </span>
      <span
        class="absolute bs-button-bg inset-0 opacity-0 transition-opacity backdrop-blur-[2px] duration-[400ms] z-[-1] group-hover:opacity-[1]"
      />
    </a>
  </div>
</form>
`;
